ui/src/sandbox.rs

```diff
@@ -167,9 +176,21 @@ impl Sandbox {
 
     pub fn execute(&self, req: &ExecuteRequest) -> Result<ExecuteResponse> {
         try!(self.write_source_code(&req.code));
-        let mut command = self.execute_command(req.channel, req.mode, req.crate_type, req.tests, req.backtrace, req.edition);
 
-        let output = try!(command.output().map_err(Error::UnableToExecuteCompiler));
+        let mut command = self.execute_command(req.channel, req.mode, req.crate_type, req.tests, req.backtrace, req.edition);
+        let output = if let Some(ref input) = req.stdin {
+            command.stdin(Stdio::piped());
+            command.stdout(Stdio::piped());
+            command.stderr(Stdio::piped());
+            let mut child = try!(command.spawn().map_err(Error::UnableToExecuteCompiler));
+            {
+                let stdin = try!(child.stdin.as_mut().ok_or(Error::UnableToOpenStdin));
+                try!(stdin.write_all(input.as_bytes()).map_err(Error::UnableToWriteInput));
```

Comment:
Here you could end by having a deadlock if input is > pipe buffer size (~65k in general), so you would wait for write and the child could wait for writing to stdout too and cause no one read the output, both child and parent would be block on writing call. That of course is a very rare scenario in this context.

The solution could be to have a second child that only write the input, so the parent can read the output, or use non blocking pipe, or use select function, etc...